File: algorithms/arrays-and-sorting/insertion-sort.cpp
// Insertion Sort Advanced Analysis
// How many shifts will it take Insertion Sort to sort an array?
//
// https://www.hackerrank.com/challenges/insertion-sort/problem
//

#include <iostream>
using namespace std;

const int N = 10000000;
int cnt[N + 1];

void add(int v)
{
    while (v <= N)
    {
        cnt[v]++;
        v += v & -v;
    }
}

int get(int v)
{
    int sum = 0;
    while (v)
    {
        sum += cnt[v];
        v -= v & -v;
    }
    return sum;
}

int main()
{
    int T;
    cin >> T;
    while (T--)
    {
        int n;
        cin >> n;

        long long changes = 0;
        for (int i = 0; i < n; i++)
        {
            int a;
            cin >> a;

            int cnt_larger = get(N) - get(a);
            changes += cnt_larger;

            add(a);
        }

        cout << changes << endl;

        for (int i = 0; i <= N; i++)
            cnt[i] = 0;
    }

    return 0;
}

File: algorithms/arrays-and-sorting/insertion-sort.py
# Insertion Sort Advanced Analysis
# How many shifts will it take Insertion Sort to sort an array?
#
# https://www.hackerrank.com/challenges/insertion-sort/problem
#

def insertionSort(arr):

      bit = [0] * 10000001
      bit_sum = 0
      count = 0

      for n in arr:
        count += bit_sum
        idx = n
        while idx:
          count -= bit[idx]
          idx -= idx & -idx

        idx = n
        while idx < 10000001:
          bit[idx] += 1
          idx += idx & -idx

        bit_sum += 1

      return count


if __name__ == "__main__":
    t = int(input().strip())
    for a0 in range(t):
        n = int(input().strip())
        arr = list(map(int, input().strip().split(' ')))
        result = insertionSort(arr)
        print(result)
